uuid.go

```diff
@@ -156,6 +158,47 @@ func (u UUID) String() string {
 	return string(buf)
 }
 
+// Format implements fmt.Formatter for UUID values.  It accepts 'h' or 'H' as a format to output only
+// the hex digits of the UUID.  Any other format specifier returns the canonical RFC-4122 representation.
+func (u UUID) Format(f fmt.State, c rune) {
+	//
+	// TODO: dylan.bourque - 2019-02-15
+	// . should we implement %v? if so, what is "a Go-syntax representation" of a UUID when using %#v?
+	//
+	switch c {
+	case 'h', 'H':
+		s := hex.EncodeToString(u.Bytes())
+		if c == 'H' {
+			s = strings.Map(toCapitalHexDigits, s)
+		}
+		_, _ = io.WriteString(f, s)
+	case 'q':
+		_, _ = io.WriteString(f, `"`+u.String()+`"`)
+	default:
+		_, _ = io.WriteString(f, u.String())
```

Comment:
I'm not convinced about this approach.

When `UUID` does not implement `fmt.Formatter`, this program fails the `go vet` check, and produces obviously wrong output: https://play.golang.org/p/stts25e3_2a

However, when `UUID` _does_ implement `fmt.Formatter`, the program passes `go vet`, and produces more plausible looking (but wrong) output: https://play.golang.org/p/hNOR123ZjkU

The behavior in the second case seems dangerous to me. Using the wrong formatting verb is a programmer error and this implementation of `fmt.Formatter` hides it from `go vet`, and swallows it when it produces output. I don't know what a good solution to this problem might be, short of trying to re-create the original format string from the `fmt.State` and calling `fmt.Sprintf` with a `[16]byte` value instead of a `UUID`, to avoid recursion.

For reference, `pkg/errors` implements `fmt.Formatter` on some of its types, and when it doesn't understand a verb, it simply produces no output, e.g. https://github.com/pkg/errors/blob/master/stack.go#L64-L84. I'm not a huge fan of this approach either, but it is at least a data point we have, if we're going to implement `fmt.Formatter` at all.

Reply:
There doesn't seem to be any way to indicate that a formatting verb is invalid/not supported, so it feels like "do nothing" is the only option.  I'm okay with that path for all of the formatting verbs that don't apply: %b, %c, %d, %o, %U, %e, %E, %f, %F, %g, %G.

Reply:
To clarify, "do nothing" means "emit no output"?

Reply:
Correct, "do nothing" would be to return an empty string for all of those verbs.

Reply:
Okay. I guess that's reasonable, and there is a precedent for it too. But my concern with respect to the fact that this conceals programmer errors such as using the wrong flags remains.

The `fmt.Formatter` implementation is opaque to static analysis tools like `go vet`. We're gaining convenience and a nice API for printing UUIDs as hexadecimal strings, but we are losing precision under the lens of static analysis tools. We should decide if this trade-off is worth making.